Src/Base/AMReX_Extension.H

```diff
@@ -151,6 +151,22 @@
 #define AMREX_FALLTHROUGH ((void)0)
 #endif
 
+#ifndef AMREX_NO_NODISCARD
+#  ifdef __has_cpp_attribute
+#    if __has_cpp_attribute(nodiscard)
+#      define AMREX_NODISCARD [[nodiscard]]
+#    elif defined(__GNUC__) || defined(__clang__)
+#      define AMREX_NODISCARD __attribute__ ((warn_unused_result))
+#    else
+#      define AMREX_NODISCARD
+#    endif
+#  elif defined(__GNUC__) || defined(__clang__)
+#    define AMREX_NODISCARD __attribute__ ((warn_unused_result))
+#  else
+#    define AMREX_NODISCARD
+#  endif
+#endif
+
```

Comment:
I like this!  It seems that the logic can be simplified with
```
#if !defined(AMREX_NO_NODISCARD) && defined(__has_cpp_attribute) && __has_cpp_attribute(nodiscard)
# 
#elif !defined(AMREX_NO_NODISCARD) && (defined(__GNU__) || defined(__clang__))
# 
#else
# 
#endif
```
I still we need to define `AMREX_NODISCARD` as empty even if `AMREX_NO_NODISCARD` is defined.

Reply:
Of course, you are right. We still need to define an empty macro.
My macro-fu is weak, so I didn't know if the preprocessor can do partial evaluation, i.e. check for the existence of a macro and use it in the and clause. I will try it!